File: src-tauri/src/commands/usage.rs
use tauri::State;

use crate::error::{AppError, AppResult};
use crate::models::usage::{DailyUsage, DayModelUsage, ModelUsage, UsageSummary, UsageSyncResult};
use crate::modules::storage::usage_repo::{self, UsageFilter};
use crate::modules::usage_local;
use crate::state::AppState;

/// 触发一次本机用量增量同步（读取 ~/.claude 与 ~/.codex 会话日志）。
///
/// 首次为全量解析，文件量大时耗时较长，故在阻塞线程池执行，避免卡死主线程。
#[tauri::command]
pub async fn sync_session_usage(state: State<'_, AppState>) -> AppResult<UsageSyncResult> {
    let pool = state.db_pool.clone();
    tauri::async_runtime::spawn_blocking(move || -> AppResult<UsageSyncResult> {
        let conn = pool.get()?;
        Ok(usage_local::sync_all(&conn))
    })
    .await
    .map_err(|e| AppError::Unknown(format!("用量同步任务失败: {e}")))?
}

/// 组装查询过滤：date 闭区间[YYYY-MM-DD]（预设周期）或 ts 毫秒闭区间（自定义时分范围）。
fn filter_of<'a>(
    start: &'a Option<String>,
    end: &'a Option<String>,
    start_ts: Option<i64>,
    end_ts: Option<i64>,
    app_type: &'a Option<String>,
) -> UsageFilter<'a> {
    UsageFilter {
        start: start.as_deref(),
        end: end.as_deref(),
        start_ts,
        end_ts,
        app_type: app_type.as_deref(),
    }
}

/// 用量总览。
#[tauri::command]
pub fn get_usage_summary(
    state: State<AppState>,
    start: Option<String>,
    end: Option<String>,
    start_ts: Option<i64>,
    end_ts: Option<i64>,
    app_type: Option<String>,
) -> AppResult<UsageSummary> {
    let conn = state.db_pool.get()?;
    usage_repo::summary(&conn, &filter_of(&start, &end, start_ts, end_ts, &app_type))
}

/// 按模型聚合用量。
#[tauri::command]
pub fn get_usage_by_model(
    state: State<AppState>,
    start: Option<String>,
    end: Option<String>,
    start_ts: Option<i64>,
    end_ts: Option<i64>,
    app_type: Option<String>,
) -> AppResult<Vec<ModelUsage>> {
    let conn = state.db_pool.get()?;
    usage_repo::by_model(&conn, &filter_of(&start, &end, start_ts, end_ts, &app_type))
}

/// 按天聚合用量。
#[tauri::command]
pub fn get_usage_by_day(
    state: State<AppState>,
    start: Option<String>,
    end: Option<String>,
    start_ts: Option<i64>,
    end_ts: Option<i64>,
    app_type: Option<String>,
) -> AppResult<Vec<DailyUsage>> {
    let conn = state.db_pool.get()?;
    usage_repo::by_day(&conn, &filter_of(&start, &end, start_ts, end_ts, &app_type))
}

/// 按本地小时聚合用量。`date` 为 `YYYY-MM-DD HH:00`；无 ts 的行不计入。
#[tauri::command]
pub fn get_usage_by_hour(
    state: State<AppState>,
    start: Option<String>,
    end: Option<String>,
    start_ts: Option<i64>,
    end_ts: Option<i64>,
    app_type: Option<String>,
) -> AppResult<Vec<DailyUsage>> {
    let conn = state.db_pool.get()?;
    usage_repo::by_hour(&conn, &filter_of(&start, &end, start_ts, end_ts, &app_type))
}

/// 按天 × 来源 × 模型聚合（多维合并表）。
#[tauri::command]
pub fn get_usage_by_day_model(
    state: State<AppState>,
    start: Option<String>,
    end: Option<String>,
    start_ts: Option<i64>,
    end_ts: Option<i64>,
    app_type: Option<String>,
) -> AppResult<Vec<DayModelUsage>> {
    let conn = state.db_pool.get()?;
    usage_repo::by_day_model(&conn, &filter_of(&start, &end, start_ts, end_ts, &app_type))
}

File: src/services/modules/usage.ts
import { request } from "../request";

export type UsageAppFilter = "all" | "claude" | "codex" | "zcode";

export interface UsageSummary {
  totalRequests: number;
  totalInputTokens: number;
  totalOutputTokens: number;
  totalCacheCreationTokens: number;
  totalCacheReadTokens: number;
}

/** 按天 × 来源 × 模型聚合（多维合并表：前端按 date 行合并展示）。 */
export interface DayModelUsage {
  date: string;
  appType: string;
  model: string;
  requests: number;
  inputTokens: number;
  outputTokens: number;
  cacheCreationTokens: number;
  cacheReadTokens: number;
}

/** 按天 × 来源聚合（热力图/趋势图数据源）。 */
export interface DailyUsage {
  date: string;
  appType: string;
  requests: number;
  inputTokens: number;
  outputTokens: number;
  cacheCreationTokens: number;
  cacheReadTokens: number;
}

export interface UsageSyncResult {
  imported: number;
  filesScanned: number;
  errors: number;
}

interface UsageFilter {
  /** date 闭区间（YYYY-MM-DD，预设周期） */
  start?: string;
  end?: string;
  /** ts 毫秒闭区间（自定义时分范围） */
  startTs?: number;
  endTs?: number;
  appType?: string;
}

const filterArgs = (f: UsageFilter) => ({
  start: f.start,
  end: f.end,
  startTs: f.startTs,
  endTs: f.endTs,
  appType: f.appType,
});

export const usageApi = {
  /** 触发本机用量增量同步。 */
  sync: () => request<UsageSyncResult>("sync_session_usage"),
  getSummary: (f: UsageFilter = {}) =>
    request<UsageSummary>("get_usage_summary", filterArgs(f)),
  /** 按天 × 来源 × 模型聚合（date 倒序、组内 token 降序）。 */
  getByDayModel: (f: UsageFilter = {}) =>
    request<DayModelUsage[]>("get_usage_by_day_model", filterArgs(f)),
  /** 按天 × 来源聚合。 */
  getByDay: (f: UsageFilter = {}) =>
    request<DailyUsage[]>("get_usage_by_day", filterArgs(f)),
  /** 按小时 × 来源聚合。`date` 为 `YYYY-MM-DD HH:00`。 */
  getByHour: (f: UsageFilter = {}) =>
    request<DailyUsage[]>("get_usage_by_hour", filterArgs(f)),
};
